x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java

```diff
@@ -333,27 +334,42 @@ private void authenticateAsync() {
-                            if (userToken != null) {
-                                writeAuthToContext(userToken.getAuthentication());
-                            } else {
-                                checkForApiKey();
-                            }
-                        }, e -> {
-                            logger.debug(new ParameterizedMessage("Failed to validate token authentication for request [{}]", request), e);
-                            if (e instanceof ElasticsearchSecurityException &&
-                                tokenService.isExpiredTokenException((ElasticsearchSecurityException) e) == false) {
-                                // intentionally ignore the returned exception; we call this primarily
-                                // for the auditing as we already have a purpose built exception
-                                request.tamperedRequest();
-                            }
-                            listener.onFailure(e);
-                        }));
+                        checkForBearerToken();
                     }
                 });
             }
         }
 
+        private void checkForBearerToken() {
+            final SecureString bearerToken = tokenService.extractBearerTokenFromHeader(threadContext);
+            serviceAccountService.tryAuthenticateBearerToken(bearerToken, nodeName, ActionListener.wrap(authentication -> {
+                if (authentication != null) {
+                    this.authenticatedBy = authentication.getAuthenticatedBy();
+                    writeAuthToContext(authentication);
+                } else {
+                    tokenService.tryAuthenticateToken(bearerToken, ActionListener.wrap(userToken -> {
+                        if (userToken != null) {
+                            writeAuthToContext(userToken.getAuthentication());
+                        } else {
+                            checkForApiKey();
+                        }
+                    }, e -> {
+                        logger.debug(new ParameterizedMessage("Failed to validate token authentication for request [{}]", request), e);
+                        if (e instanceof ElasticsearchSecurityException
+                            && false == tokenService.isExpiredTokenException((ElasticsearchSecurityException) e)) {
+                            // intentionally ignore the returned exception; we call this primarily
+                            // for the auditing as we already have a purpose built exception
+                            request.tamperedRequest();
+                        }
+                        listener.onFailure(e);
+                    }));
+                }
+            }, e -> {
+                logger.debug(new ParameterizedMessage("Failed to validate service account token for request [{}]", request), e);
+                listener.onFailure(request.exceptionProcessingRequest(e, null));
```

Comment:
I think we need to revisit the `null` here - we should be auditing with an AuthenticationToken since we have a credential.

Reply:
Good point. I made `ServiceAccountToken` implement `AuthenticationToken` and used it here. For this purpose, I also altered (reverted) the code to check for the token in `AuthenticationService` and this is actually what your poc code was suggesting.